[FEAT] 대결 신청 API 기능 추가 + 통합테스트 작성

park-se-jun opened the pull request:
## PR Description 🗒️
- 대결신청 API 추가
- mockMvc로 통합테스트 작성
- 애매한 부분 에 대해서 유닛테스트 작성
## 추가/변경 사항 및 설명 📝
- 막은 오류사항
  - 같은 노래는 배틀 신청을 못한다
  - 배틀신청할 수 있는 postId 가 아니면 배틀신청을 못한다.
  - 사용자 소유의 post가 아니면 배틀 신청을 못한다.
  - post의 genre가 다르면 배틀 신청을 못한다.
  - 대결권이 부족하면 배틀 신청을 못한다.
  - 사용자 소유의 post에 배틀 신청을 못한다.
## To Reviewers 🙏

## Issue close ✂️
close #92

park-se-jun committed:
[CHORE] local용 db를 실행 시키는 docker compose 작성

park-se-jun committed:
[FEAT] flyway 적용 및 로컬 환경에서의 db 연결

park-se-jun committed:
[CHORE] application.yml 환경별 분리
- test 환경의 경우 h2사용하도록 분리
- local 환경의 경우 localhost:3306을 사용하도록 분리
- prod 환경일 경우 아직 작성 못함

park-se-jun committed:
[CHORE] prod 관련 applicatoin.yml 작성 및 cd 수정
- application-prod.yml 작성
- deploy.sh 에 profile 관련 옵션 추가
- deploy workflow에 prod의 db 관련 정보 설정

park-se-jun committed:
Merge pull request #57 from prgrms-web-devcourse/feature_#41

[FEAT] 플라이웨이 적용, test db docker compose 작성, 환경별 application.yml 분리

park-se-jun committed:
[FEAT] cors config 작성

- 아직 동작 x

park-se-jun committed:
[FEAT] basic security config 작성

- 아직 동작 x

park-se-jun committed:
Merge branch 'develop' into sejun

park-se-jun committed:
[CHORE] 임시로 pr과 deploy.yml 연결

park-se-jun committed:
[CHORE] pr 과 deploy.yml 연결 제거, 테스트용 db pr-check.yml에 추가

park-se-jun committed:
[CHORE] test시 test profile을 이용하도록 build 설정
- docker compose 관련 워크플로 삭제
- test 시 test profile 을 이용하도록 변경

park-se-jun committed:
[CHORE] pr에 deploy.yml 연결
- deploy.yml 트리거로 pr 연결
- pr-check 의 이름을 pr-check로 변경(원래 develop이었음)

park-se-jun committed:
[CHORE] shell script 수정 - profile을 prod로

park-se-jun committed:
[CHORE] 배포는 되지만 실행이 되지 않는 문제
- application-prod.yml 에 env를 file path로 import 하는데, file path는 커맨드 실행하는데를 기준으로 해서 배포시 잡지 못하는 문제가 있었음
- deploy.sh 에 -Dspring.config.import 옵션으로 덮어쓰기

park-se-jun committed:
[CHORE] spring.config.import 를 classpath로 변경한거에서 filepath로 되돌리기
- shell 파일 수정을 통해, 절대경로로 prod_db_info를 받게끔 했는데 application야믈 에는 classpath로 되어 잇었

park-se-jun committed:
[CHORE] deploy.yml에 임시로 pr 트리거 걸어 뒀던 것 제

park-se-jun committed:
[FEAT] 가장 기본적인 oauth2설정

park-se-jun committed:
[FIX] 오타 수정

park-se-jun committed:
[FEAT] 중간 때 참조 했던 코드들 붙여넣기( 이걸로 안할 수도 있음)

park-se-jun committed:
[CHORE] OAuth2 login 시 필요한 application.yml 수정 및 그에 해당하는 properties 생성
- client-id, client-secret 같이 필요한 정보들 은닉화 및 추가
- 그외 사용하는 정보 application에 추가
- 추가한 정보 properties로 생성

park-se-jun committed:
[DEL] 지원하지 않는  OAuth2 Userinfo 삭제

park-se-jun committed:
[DEL] Member의 refreshToken 관련 assertion 삭제
- refreshToken은 처음에 null 일 수 있음

park-se-jun committed:
[FEAT] TokenProvider를 우리의 코드에 맞게 수정

park-se-jun committed:
[FEAT] SecurityConfig수정
- h2-consle 관련 설정
- authorizationRequestRepository 설정
- SucessHandler 설정
- TokenFilter 설정

park-se-jun committed:
[FEAT] socialinfo assertion 잘못 되 있던거 수정

park-se-jun committed:
[FEAT] OAuth2 인증(Authentication) 성공시 DB 저장 로직 구현
- Authentication 성공시 MemberRepository에서 찾아봄
- 없다면 DB에 생성
- 있다면, 있는거 떠내서 씀
- 해당하는 userId로 AccessToken 만듦

park-se-jun committed:
[FEAT] 요청이 들어왔을 때 AccessToken을 검증하는 필터를 코드 상황에 맞게 수정

park-se-jun committed:
[FEAT] CORS 설정 적용되도록 변경
- @Configuration 어노테이셔 추가
- allowedOrigins를 @Value로 꽂아줌

park-se-jun committed:
[FEAT] 리프레스 토큰 관련 코드 추가

park-se-jun committed:
CHORE] 혹시 모를 front, back 주소 관련 application yml 수정

park-se-jun committed:
[CHORE] sensitive.env 추가에 따른 workflow 수정

park-se-jun committed:
sejun과 feature_#33의  conflict 해소 및 merge

park-se-jun committed:
Merge pull request #84 from prgrms-web-devcourse/feature_#33

[FEAT] OAuth2로 로그인 기능 추가 및 설정

park-se-jun committed:
[FIX] develop 과 sejun merge 해소

park-se-jun committed:
[FIX] code format 맞추기

park-se-jun committed:
[CHORE] code format 맞추기

park-se-jun committed:
[DEL] refreshToken 관련 검증 테스트 삭제

park-se-jun committed:
[FIX] 모든 요청에 대해 인증을 요구 하지 않도록 변경

park-se-jun committed:
[FIX] Genre 관련 controller 테스트가 Junit4 기반으로 되어 있던거 수정
- Junit4 기반으로 되어 있어서 테스트 인식을 하지 못하던 오류 수정

park-se-jun committed:
[FIX] Security 적용 과정에서 테스트가 실패하는 부분 수정
- TokenAuthenticationFilter를 스캔 대상에서 제외
- 위의 필터를 생성하기 위해선 TokenProvider와 AppProperties가 필요한데 이를 넣어주기 보단 필터 자체를 스캔 안하도록 변경

park-se-jun committed:
[FIX] 아직 테스트 하지 못한코드들을 jacoco의 exclude로 등록
- security 코드 전반
- security 코드에서 사용하는 util

park-se-jun committed:
[FEAT] OAuth2AuthenticationException 핸들링 포인트 추가

park-se-jun committed:
[FEAT] assignRefreshToken의 에러 처리시 ExceptionMessage로 처리

park-se-jun committed:
[CHORE] 후에 좀더 의미있는 예외보 바꾸기 위해 todo 추가

park-se-jun committed:
[ADD] develop으로부터 merge

park-se-jun committed:
[FEAT] createBattle 컨트롤러 코드 작성, 성공 테스트 코드 작성

- BattleController 에 createBattle 메소드 작성
- BattleService의 base code  작성
- createBattle에 대한 controller test 작성(restdocs)
- ResponseMessage에 success create battle 추

park-se-jun committed:
[REFACTOR] BattleService 위치 변경

park-se-jun committed:
[FEAT] Post의 Id와, 배틀 가능 여부에 따라 Post를 가져오는 메소드 추가

park-se-jun committed:
[FEAT] BattleRepository 추

park-se-jun committed:
[FEAT] createBattle기능 구현
- 멤버가 소유한 post인지 검증
- 입력받은 PostId로 검색된 post가 배틀 가능하면서 존재하는지 검증
- 대결받는 포스트가 멤버가 가지는 포스트가 아닌지 검증
- 들어온 두 포스트의 장르가 일치하는지 검증
- 각각의 에러 상황에 맞는 ExceptionMessage 작성

park-se-jun committed:
[FEAT] createBattle 대결권 줄이기 로직 추가

park-se-jun committed:
[TEST] createBattle servie 테스트 뼈대 작성

park-se-jun committed:
[FIX] develop으로 부터 merge

park-se-jun committed:
[DEL] createBattle 단위테스트 뼈대 일단 제거

park-se-jun committed:
[DEL] IllegalStateException 핸들러 추가

park-se-jun committed:
[ADD] BattleService의 Repository가 final이 아니던 것 수정

park-se-jun committed:
[FIX] 배틀 신청하는 두 포스트의 MusicId가 같은것을 validation

park-se-jun committed:
[FIX] ExceptionMessage 수정

park-se-jun committed:
[FIX] Create Battle MockMvc를 이용한 통합 테스트 작

thguss commented on src/main/java/com/example/demo/controller/battle/BattleCreateRequestDto.java:
```diff
@@ -0,0 +1,12 @@
+@Builder
+public record BattleCreateRequestDto(
+	@NotNull long challengedPostId,
+	@NotNull long challengingPostId
```
long이 원시타입이기 때문에 Null이 들어올 수가 없어요.
- @NotNull 을 제거하거나
- long 원시타입을 Long 래퍼타입으로 바꾸거나

둘 중 한 방식으로 진행해도 좋을 것 같아요~!

park-se-jun replied:
아래방식으로다가

thguss commented on src/main/java/com/example/demo/model/member/Member.java:
```diff
@@ -130,5 +130,9 @@ public void addOneChallengeTicket() {
+	public void subtractCountOfChallengeTicket() {
+		this.countOfChallengeTicket--;
+	}
 	// TODO: 2023-02-23 member가 특정 배틀에 투표한적 있는지 확인하는 메소드(할수 없을 것 같기도)
```
이거 완료됐나요?

thguss commented on src/main/java/com/example/demo/service/BattleService.java:
```diff
@@ -17,7 +28,84 @@
+	private void validSameMusic(Post challengingPost, Post challengedPost) {
+		String challenedPostMusicId = challengedPost.getMusic().getMusicId();
+		String challengingPostMusicId = challengingPost.getMusic().getMusicId();
+		if (challengingPostMusicId == challenedPostMusicId) {
```
`==`  표시 대신 Objects.equals 는 어때요?

twotwobread replied:
String이면 == 하면 더욱 안될것 같아요!

thguss commented on src/main/java/com/example/demo/service/BattleService.java:
```diff
@@ -17,7 +28,84 @@
+	}
+
+	private Genre validGenre(Post challengingPost, Post challengedPost) {
+		if (challengingPost.getMusic().getGenre() != challengedPost.getMusic().getGenre()) {
```
이거 또한 Objects.equals

thguss commented on src/test/java/com/example/demo/repository/PostRepositoryTest.java:
```diff
@@ -4,18 +4,23 @@
+import lombok.extern.slf4j.Slf4j;
+
 @DataJpaTest
+@Slf4j
```
로그를 찍으신 이유가 있을까요

thguss commented on src/test/java/com/example/demo/service/BattleServiceTest.java:
```diff
@@ -73,6 +74,43 @@ class BattleServiceTest {
+		@Test
+		void 실패_createBattle_배틀가능한_postId를_찾을_수_없음_EntityNotFoundException() {
+
+		}
```
?

thguss reviewed (commented):
수고하셨습니다~!

twotwobread commented on src/main/java/com/example/demo/controller/battle/BattleCreateRequestDto.java:
```diff
@@ -0,0 +1,12 @@
+
+@Builder
+public record BattleCreateRequestDto(
+	@NotNull long challengedPostId,
```
primitive면 NotNull이 걸릴수가 있나요?? 요청에서 primitive를 사용해도 상관없나요?? (이건 진짜 몰라서,,,)

twotwobread commented on src/main/java/com/example/demo/service/BattleService.java:
```diff
@@ -17,7 +28,84 @@
+			.status(BattleStatus.PROGRESS)
+			.build();
+		battleRepository.save(newBattle);
+		memberByPrincipal.subtractCountOfChallengeTicket();
```
여기서 대결 측에서 Post와의 연관관계는 대결을 생성하면서 연결되지만 Post 측에서 Battle과의 연관관계를 설정해주는 부분이 있어야 할 것 같고 연관관계 편의 메소드 하나 있어야 할 것 같아요!!

park-se-jun replied:
하긴 했는데 제대로 됬는지 확인 부탁드립니다.

twotwobread commented on src/main/resources/db/migration/V2.2__Delete_Battle_Member_relation.sql:
```diff
@@ -1,4 +1,6 @@
+FOREIGN KEY FKfhohk3campcgr4i0738qr95n7;
 ALTER TABLE battle
-    DROP COLUMN member_id;
+DROP
```
요렇게 하는게 문법 상 맞는 걸까요? 다음에 바꾸게 되면 참고하려구요!

twotwobread commented on src/test/java/com/example/demo/controller/battle/BattleControllerTest.java:
```diff
@@ -0,0 +1,421 @@
+
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
```
이거 쓰나요??

twotwobread commented on src/test/java/com/example/demo/controller/battle/BattleControllerTest.java:
```diff
@@ -0,0 +1,421 @@
+@Slf4j
+@WithMockUser
+@ExtendWith(RestDocumentationExtension.class)
+// @WebMvcTest(value = BattleController.class)
```
주석 삭제 해주세요!!

twotwobread commented on src/test/java/com/example/demo/controller/battle/BattleControllerTest.java:
```diff
@@ -0,0 +1,421 @@
+				.content(mapper.writeValueAsString(battleCreateRequestDto))
+				.header("Authorization", "Bearer {AccessToken}")
+				.with(csrf()
+				)
```
요거 올리면 좋을듯!

twotwobread reviewed (approved):
approve

park-se-jun committed:
[FIX] Dto id wrapper타입으로 변경

park-se-jun committed:
[FIX] Validation시 string,Genre에 대해 == 로 비교하던 것을 equals 로 비교하도록 변경

park-se-jun committed:
[FIX] 진행하지 않은 service 테스트 삭제

park-se-jun committed:
[FIX] 테스트에 필요하지 않은 log 삭제

park-se-jun committed:
[DEL] 테스트에 필요하지 않은 로그 제거, 주석 제거

park-se-jun committed:
[FIX] 괄호 줄바꿈 변경

park-se-jun committed:
[FIX] 배틀 생성시 post와의 연관관계 설정 편의 메소드 추가

thguss reviewed (approved):
수고하셨습니다~! 충돌 해결하고 머지 부탁드려요.

park-se-jun committed:
[FIX] develop과의 충돌 해결 및 merge
- MemberServiceTest 의 240행 메소드 createBattle에서 post에서 challengedBattles를 붙여주는걸 제거(연관관계 편의 메소드가 설정되었기 때문)
- 자잘한 파일 위치 변
